src/mail.sh

```diff
@@ -0,0 +1,125 @@
+# This file handles all the interactions of git send-email. It handles its
+# configuration proccess to facilitate the sending of patches via e-mail.
+
+include "$KW_LIB_DIR/kwlib.sh"
+include "$KW_LIB_DIR/kwio.sh"
+include "$KW_LIB_DIR/kw_config_loader.sh"
+
+# Hash containing user options
+declare -gA options_values
+
+function kw_mail()
+{
+  parse_mail_options "$@"
+  if [[ "$?" -gt 0 ]]; then
+    complain "${options_values['ERROR']}"
+    return 22 # EINVAL
+  fi
+
+  if [[ "${options_values['SETUP']}" == 1 ]]; then
+    mail_setup
+    exit
+  fi
+
+  return 0
+}
+
+# This function deals with configuring the necessary options to use
+# the git send-email tool to send patches by email
+function mail_setup()
+{
+  if ! git rev-parse --is-inside-work-tree &> /dev/null && [[ -z "${options_values['GLOBAL']}" ]]; then
+    complain 'Not in a git repository, aborting setup!'
+    say 'To apply settings globally rerun with "--global" flag.'
+    exit 1 # EPERM
+  fi
+
+  for option in "${!options_values[@]}"; do
+    if [[ "$option" =~ ^SMTP.* ]] && [[ -n "${options_values[$option]}" ]]; then
+      validate_encryption "$option"
+      # TODO change to eval
+      echo "git config ${options_values['GLOBAL']}${options_values[$option]}"
+    fi
+  done
+
+  return 0
+}
+
+# encryption must be 'ssl' or 'tls' this validates this option
+function validate_encryption()
+{
+  if [[ "$1" =~ .*ENCRYPTION ]]; then
+    case "${options_values["$1"]: -3}" in
+      ssl | tls)
+        return 0
+        ;;
+      *)
+        say 'Invalid encryption, must be "ssl" or "tls"'
+        say 'Defaulting to plain smtp'
+        options_values["$1"]="sendemail.${1,,} smtp"
+        return 0
+        ;;
+    esac
+  fi
+}
+
+function parse_mail_options()
+{
+  local short_options='h,c,'
+  local long_options='help,setup,global,'
+  long_options+='email:,name:,'
+  long_options+='smtpuser:,smtpencryption:,smtpserver:,smtpserverport:,smtppass:,'
+
+  options="$(kw_parse "$short_options" "$long_options" "$@")"
+  if [[ "$?" != 0 ]]; then
+    options_values['ERROR']="$(kw_parse_get_errors 'kw mail' "$short_options" \
+      "$long_options" "$@")"
+    return 22 # EINVAL
+  fi
+
+  # Default values
+  options_values['SETUP']=0
+  options_values['GLOBAL']=''
+
+  eval "set -- $options"
+
+  while [[ "$#" -gt 0 ]]; do
+    case "$1" in
+      --help | -h)
+        mail_help "$1"
+        exit
+        ;;
+      --setup | -c)
+        options_values['SETUP']=1
+        shift
+        ;;
+      --smtp*)
+        local option="${1#--}"
+        options_values["${option^^}"]="sendemail.$option $2"
```

Comment:
If I'm not wrong, `^^` capitalize `$option`, right? How about add `upcase_str` in `kw_string.sh` and use it here?

Reply:
Also, don't we need to validate the `$option`? We might try to set something invalid. Maybe you can create an array with all the possible valid values for `sendemail.SOMETHING` and check if `$option` matches one of that options.

Reply:
Option validation happens in the `kw_parse` function, since it will return an error if the options are not listed in either long_options or short_options.

Reply:
Yeah, but from what I can see here, there is no validation for `$option`. For example, the user passes `--smpt-lalaxpto`, it looks like that we will accept it as a valid option. I suggest you check `--smtp$option` in the easy stages; you can do it by creating an array with valid options for `smtp`.